File: examples/ugly.cs
using System;
using System.Collections;

public class App
{
	public static long value(long x, long y, long z)
	{
		return (long)(Math.Pow(2, x)*Math.Pow(3, y)*Math.Pow(5, z));
	}
	
	public static long ugly(int max)
	{		
		ArrayList uglies = new ArrayList();
		long counter = 1;
		Hashtable dict = new Hashtable();
		dict[counter] = new long[] { 0, 0, 0 };		
		
		while (uglies.Count < max)
		{	
			uglies.Add(counter);
			long[] array = (long[])dict[counter];
			long x = array[0];
			long y = array[1];
			long z = array[2];			
			
			dict[value(x+1, y, z)] = new long[] { x+1, y, z };
			dict[value(x, y+1, z)] = new long[] { x, y+1, z };
			dict[value(x, y, z+1)] = new long[] { x, y, z+1 };
			
			dict.Remove(counter);
			
			ArrayList keys = new ArrayList(dict.Keys);
			keys.Sort();
			
			counter = (long)keys[0];
		}
	
		return (long)uglies[uglies.Count-1];
	}
	
	public static void Main()
	{
		int iter = 1500;
		DateTime start = DateTime.Now;
		long uvalue = 0;
		for (int i=0; i<10; ++i)
		{
			uvalue = ugly(iter);
		}
		DateTime end = DateTime.Now;
		Console.WriteLine("{0} ugly value = {1} in {2}ms", iter, uvalue, (end-start).TotalMilliseconds);
	}
}

File: examples/ugly.py
from time import time

def value(x, y, z):
	return 2**x * 3**y * 5**z

def ugly(maxValue):
	uglies = []
	counter = 1
	d = {1 : (0, 0, 0)}

	while len(uglies) < maxValue:	
		uglies.append(counter)
		x, y, z = d[counter]
		
		d[value(x+1, y, z)] = (x+1, y, z)
		d[value(x, y+1, z)] = (x, y+1, z)
		d[value(x, y, z+1)] = (x, y, z+1)
		
		del d[counter]
		
		keys = d.keys()
		keys.sort()		
		counter = keys[0]
	
	return uglies[-1]
	
iterations = 1500

start = time()
for i in range(10):
	uvalue = ugly(iterations)
end = time()
print iterations, "ugly value =", uvalue, "in", (end - start)*1000, "ms" 

